src/Microsoft.OData.Core/Json/JsonReader.cs

```diff
@@ -921,59 +927,55 @@ private string ConsumeTokenToString(int characterCount)
         /// all indeces to the characterBuffer are invalid except for tokenStartIndex.</remarks>
         private bool ReadInput()
         {
-            Debug.Assert(this.storedCharacterCount <= this.characterBuffer.Length, "We can only store as many characters as fit into our buffer.");
             Debug.Assert(this.tokenStartIndex >= 0 && this.tokenStartIndex <= this.storedCharacterCount, "The token start is out of stored characters range.");
 
             if (this.endOfInputReached)
             {
                 return false;
             }
 
+            // initialze the buffer
+            if (this.characterBuffer == null)
+            {
+                this.characterBuffer = BufferUtils.RentFromBuffer(ArrayPool, InitialCharacterBufferSize);
+            }
+            Debug.Assert(this.storedCharacterCount <= this.characterBuffer.Length, "We can only store as many characters as fit into our buffer.");
+
             // We need to make sure we have more room in the buffer to read characters into
             if (this.storedCharacterCount == this.characterBuffer.Length)
             {
                 // No more room in the buffer, move or grow the buffer.
                 if (this.tokenStartIndex == this.storedCharacterCount)
                 {

```

Comment:
Should we move this check before 945?  That is, if the current token ends at the last character, shouldn't we always restart at the beginning of the buffer?

Reply:
Either is ok for me. I moved it up.

---
In reply to: [253984362](https://github.com/OData/odata.net/pull/1388#discussion_r253984362) [](ancestors = 253984362)